plugins/woocommerce/includes/class-woocommerce.php

```diff
@@ -584,7 +585,10 @@ public function includes() {
 			include_once WC_ABSPATH . 'includes/admin/class-wc-admin.php';
 		}
 
-		if ( $this->is_request( 'frontend' ) ) {
+		// We load frontend includes in the post editor, because they may be invoked via pre-loading of blocks.
+		$in_post_editor = doing_action( 'load-post.php' ) || doing_action( 'load-post-new.php' );
+
+		if ( $this->is_request( 'frontend' ) || $in_post_editor ) {
```

Comment:
If you add the `rest-api` case mentioned above, you could use that here instead of `$in_post_editor`. Rationale being that there could be other non-frontend cases besides the post editor where a REST API request could potentially attempt to render a shortcode.

Reply:
On the other hand, `frontend_includes` loads 12 more files, and I'm not sure whether that would appreciably impact REST API performance or not...

Reply:
Just going to sit here and talk to myself while doing this PR review... 😅 

I think I would err on the side of preventing fatals, and if loading those extra files during every REST API request does end up having some kind of noticeable performance impact, we could devise a workaround for that in a separate PR.

Reply:
Updated in https://github.com/woocommerce/woocommerce/pull/40648/commits/b43a8cfc47d038b0c0c87f4873f69a70b0fc36cb.

- Though I decided in https://github.com/woocommerce/woocommerce/pull/40648/commits/9425dd049d81d38c76a18618144c7f1487852b77 *not* to support a `'rest-api'` case, I have added a call to `is_rest_api_request` which amounts to the same thing.
- We cannot strip the `$in_post_editor` check. To see this element of the problem, go ahead and create a page or post with the test shortcode, publish, then remove this test and reload the editor (fatal error).

Reply:
Regarding performance, we're chiefly loading class and function definitions and are not executing bodies of logic. That should be extremely low overhead (especially with opcode caching). That said, there are indeed some side effects as a result of including those files (ie, where we call `add_action()` without anything being invoked) ... even so, there are very few of these and I still expect overhead to be very low.